Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm

```diff
@@ -315,7 +315,68 @@
 
 void WebExtensionAPIBookmarks::update(NSString *bookmarkIdentifier, NSDictionary *changes, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
 {
-    callback->reportError(@"unimplemented");
+    if (!bookmarkIdentifier || ![bookmarkIdentifier isKindOfClass:[NSString class]] || !bookmarkIdentifier.length) {
+        if (outExceptionString)
+            *outExceptionString = toErrorString(nullString(), @"bookmarkID", @"property must be a non-empty string.").createNSString().autorelease();
+        return;
+    }
+
+    if (!changes || ![changes isKindOfClass:[NSDictionary class]]) {
+        if (outExceptionString)
+            *outExceptionString = toErrorString(nullString(), @"changes", @"property must be an object (dictionary).").createNSString().autorelease();
+        return;
+    }
+
+    static NSDictionary<NSString *, id> *types = @{
+        urlKey: NSString.class,
+        titleKey: NSString.class
+    };
+
+    if (!validateDictionary(changes, @"changes", nil, types, outExceptionString))
+        return;
+
+    std::optional<WTF::String> newTitle;
+    std::optional<WTF::String> newURL;
+
+    id titleObj = changes[@"title"];
+    if (titleObj) {
+        if (![titleObj isKindOfClass:[NSString class]]) {
+            if (outExceptionString)
+                *outExceptionString = toErrorString(nullString(), @"title", @"property must be a string.").createNSString().autorelease();
+            return;
+        }
+        newTitle = dynamic_objc_cast<NSString>(titleObj);
+    }
+
+    id urlStringObj = changes[@"url"];
+    if (urlStringObj) {
+        if (![urlStringObj isKindOfClass:[NSString class]]) {
+            if (outExceptionString)
+                *outExceptionString = toErrorString(nullString(), @"url", @"property must be a string.").createNSString().autorelease();
+            return;
+        }
```

Comment:
Ditto about getting rid of this as well.

```
    static NSDictionary<NSString *, id> *types = @{
        urlKey: NSString.class,
        titleKey: NSString.class
    };

    if (!validateDictionary(changes, @"changes", nil, types, outExceptionString))
        return;
```

This should make sure that title and url are `NSString` objects.